cmd/promoted-image-governor/main.go

```diff
@@ -303,6 +309,80 @@ func isMirroredFromOCP(tag api.ImageStreamTagReference, refs []ImageStreamRef) b
 	return false
 }
 
+func deleteTagsOnBuildFarm(ctx context.Context, appCIClient ctrlruntimeclient.Client, buildClusterClients map[string]ctrlruntimeclient.Client, imageStreamsWithPromotedTags map[ctrlruntimeclient.ObjectKey]interface{}, dryRun bool) error {
+	var errs []error
+	for streamKey := range imageStreamsWithPromotedTags {
+		for cluster, client := range buildClusterClients {
+			imageStream := &imagev1.ImageStream{}
+			if err := client.Get(ctx, streamKey, imageStream); err != nil {
+				if !kerrors.IsNotFound(err) {
+					errs = append(errs, fmt.Errorf("could not get image stream %s in namespace %s on cluster %s: %w", streamKey.Name, streamKey.Namespace, cluster, err))
+				} else {
+					logrus.WithField("cluster", cluster).WithField("streamKey", streamKey).Debug("image stream not found")
+				}
+				continue
+			}
+
+			appCIImageStream := &imagev1.ImageStream{}
+			if err := appCIClient.Get(ctx, streamKey, appCIImageStream); err != nil {
+				if !kerrors.IsNotFound(err) {
+					errs = append(errs, fmt.Errorf("could not get image stream %s in namespace %s on cluster %s: %w", streamKey.Name, streamKey.Namespace, appCIContextName, err))
+				} else {
+					logrus.WithField("cluster", cluster).WithField("streamKey", streamKey).Info("deleting image stream on build farm")
+					if dryRun {
+						continue
+					}
+					if err := client.Delete(ctx, imageStream); err != nil {
+						if !kerrors.IsNotFound(err) {
+							errs = append(errs, fmt.Errorf("could not delete image stream %s in namespace %s on cluster %s", streamKey.Name, streamKey.Namespace, cluster))
+						} else {
+							logrus.WithField("cluster", cluster).WithField("streamKey", streamKey).Debug("image stream not found upon deleting")
+						}
+						continue
+					}
+					logrus.WithField("cluster", cluster).WithField("streamKey", streamKey).Info("image stream is deleted")
+				}
+				continue
+			}
+
+			tags := sets.NewString()
+			for _, tag := range imageStream.Status.Tags {
+				tags.Insert(tag.Tag)
+			}
+
+			for _, tag := range appCIImageStream.Status.Tags {
```

Comment:
nit: this might be a little more legibile if you create two sets and do a `Difference()`

Reply:
Updated.